govcd/vm.go

```diff
@@ -850,3 +851,187 @@ func (vm *VM) SetGuestCustomizationSection(guestCustomizationSection *types.Gues
+		}
+	}
+
+	if vm.VM.VmSpecSection.DiskSection == nil || vm.VM.VmSpecSection.DiskSection.DiskSettings == nil ||
+		len(vm.VM.VmSpecSection.DiskSection.DiskSettings) == 0 {
+		return nil, fmt.Errorf("cannot get internal disk - VM doesn't have internal disks")
+	}
+
+	for _, diskSetting := range vm.VM.VmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.DiskId == diskId {
+			return diskSetting, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// DeleteInternalDisk delete disk using provided disk ID.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) DeleteInternalDisk(diskId string) error {
+	err := vm.Refresh()
+	if err != nil {
+		return fmt.Errorf("error refreshing VM: %s", err)
+	}
+
+	diskSettings := vm.VM.VmSpecSection.DiskSection.DiskSettings
+	if diskSettings == nil {
+		diskSettings = []*types.DiskSettings{}
+	}
+
+	diskPlacement := -1
+	for i, diskSetting := range vm.VM.VmSpecSection.DiskSection.DiskSettings {
+		if diskSetting.DiskId == diskId {
+			diskPlacement = i
+		}
+	}
+
+	if diskPlacement == -1 {
+		return ErrorEntityNotFound
+	}
+
+	// remove disk from slice
+	diskSettings = append(diskSettings[:diskPlacement], diskSettings[diskPlacement+1:]...)
+
+	vmSpecSection := vm.VM.VmSpecSection
+	vmSpecSection.DiskSection.DiskSettings = diskSettings
+
+	_, err = vm.UpdateInternalDisks(vmSpecSection)
+	if err != nil {
+		return fmt.Errorf("error deleting VM %s internal disk %s: %s", vm.VM.Name, diskId, err)
+	}
+
+	return nil
+}
+
+// UpdateInternalDisks applies disks configuration for the VM.
+// types.VmSpecSection requires consist of all disk entities which exist and not updated,
+// as also new ones or changed ones. Returns new disk ID and error.
+// Runs synchronously, VM is ready for another operation after this function returns.
+func (vm *VM) UpdateInternalDisks(disksSettingToUpdate *types.VmSpecSection) (*types.VmSpecSection, error) {
+	if vm.VM.HREF == "" {
+		return nil, fmt.Errorf("cannot update internal disks - VM HREF is unset")
+	}
+
+	task, err := vm.UpdateInternalDisksAsync(disksSettingToUpdate)
+	if err != nil {
+		return nil, err
+	}
+	err = task.WaitTaskCompletion()
+	if err != nil {
+		return nil, fmt.Errorf("error waiting for task completion after internal disks update for VM %s: %s", vm.VM.Name, err)
+	}
+	err = vm.Refresh()
+	if err != nil {
+		return nil, fmt.Errorf("error refreshing VM %s: %s", vm.VM.Name, err)
+	}
+	return vm.VM.VmSpecSection, nil
+}
+
+// UpdateInternalDisksAsync applies disks configuration and return task or err
+// types.VmSpecSection requires consist of all disk entities which exist and not updated,
+// as also new ones or changed ones.
+func (vm *VM) UpdateInternalDisksAsync(disksSettingToUpdate *types.VmSpecSection) (Task, error) {
+	if vm.VM.HREF == "" {
+		return Task{}, fmt.Errorf("cannot update disks, VM HREF is unset")
+	}
+
+	disksSettingToUpdate.Info = "Virtual hardware requirements (simplified)"
```

Comment:
1. Where does this hard coded value come from?
2. And where will it go?
3. Does it have to be exactly like that or can it be customized?

Reply:
1. It's from: 
```
 <VmSpecSection>
            <ovf:Info>Virtual hardware requirements (simplified)</ovf:Info>
            <OsType>centos7_64Guest</OsType>
            <NumCpus>4</NumCpus>
            <NumCoresPerSocket>1</NumCoresPerSocket>
            <CpuResourceMhz>
```
2. Goes with POST (updating VM internal disk)
3. When it's not there, API complains

Reply:
I rechecked with a few version and removed it. Seems previously missing this field show errors due other things.